# Flight Line Optimization

**Purpose:** This notebook demonstrates how to use HyPlan's flight optimizer to automatically order flight lines and schedule them across sorties, minimizing total transit time while respecting aircraft endurance, refueling needs, and multi-day scheduling constraints. Optimization is beneficial whenever you have more than a handful of flight lines -- manual ordering becomes impractical and suboptimal as the number of lines grows, and the optimizer ensures that transit time between lines is minimized so that more of the flight budget is spent collecting data.

**Metadata:**
- **Audience:** Intermediate
- **Runtime:** ~1-2 minutes
- **Requires internet:** Yes (airport database download on first run)
- **Credentials required:** None
- **Optional dependencies:** `folium` (for interactive map)
- **Uses example data:** No (generates flight lines programmatically over a study area polygon)

**What You Will Learn:**
- How to build a transit-time graph connecting flight line endpoints and airports
- How the greedy nearest-neighbor optimizer orders flight lines to minimize transit time
- How endurance constraints trigger automatic refueling stops at nearby airports
- How multi-day scheduling distributes flight lines across multiple sorties
- How to feed the optimized sequence into `compute_flight_plan()` for detailed timing and geometry
- How to visualize optimized routes on static and interactive maps

### How the Optimizer Works

The optimizer solves the **Traveling Salesman Problem** variant for flight planning in two steps:

1. **`build_graph`** -- Constructs a directed graph connecting all flight line endpoints and airports, with edges weighted by transit time (computed from aircraft speed and Dubins path distances).
2. **`greedy_optimize`** -- Traverses the graph using a nearest-neighbor heuristic, inserting refuel stops and day breaks as needed.

The nearest-neighbor heuristic is not globally optimal, but it produces good solutions quickly and handles real-world constraints (endurance, refueling, multi-day scheduling) that exact solvers cannot easily accommodate.

In [1]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

from hyplan import setup_logging
setup_logging()  # INFO-level handler on the 'hyplan' logger; idempotent

import numpy as np
import matplotlib.pyplot as plt
import geopandas as gpd
from shapely.geometry import Polygon

from hyplan import (
    FlightLine, AVIRIS3,
    KingAirB200,
    Airport, initialize_data,
    box_around_polygon,
    build_graph, greedy_optimize,
    compute_flight_plan, plot_flight_plan, plot_altitude_trajectory,
    map_flight_lines,
    ureg,
)

## Setup: Flight Lines and Airports

Before optimization, we need three ingredients: a set of flight lines to fly, an aircraft with known performance, and airports for departure/return/refueling. Here we generate 16 parallel flight lines over the Channel Islands study area using `box_around_polygon()`, which automatically computes line spacing based on the sensor's swath width and the requested overlap.

In [2]:
# Initialize airport database
initialize_data(countries=["US"])

# Define sensor and aircraft
sensor = AVIRIS3()
aircraft = KingAirB200()
altitude = ureg.Quantity(20000, "feet")

# Study area polygon (larger region around Channel Islands)
study_area = Polygon([
    (-120.50, 34.20), (-120.40, 34.00), (-120.10, 33.85),
    (-119.70, 33.80), (-119.40, 33.90), (-119.30, 34.05),
    (-119.40, 34.20), (-119.70, 34.30), (-120.10, 34.35),
    (-120.40, 34.25), (-120.50, 34.20),
])

# Generate flight lines
flight_lines = box_around_polygon(
    instrument=sensor,
    altitude_msl=altitude,
    polygon=study_area,
    box_name="ISL",
    overlap=20.0,
    clip_to_polygon=False,
)

# Select airports
departure = Airport("KSBA")  # Santa Barbara
refuel = Airport("KOXR")    # Oxnard
airports = [departure, refuel]

print(f"{len(flight_lines)} flight lines generated")
print(f"Departure: {departure.name} ({departure.icao_code})")
print(f"Refuel option: {refuel.name} ({refuel.icao_code})")

2026-05-05 12:45:43,028 hyplan.download INFO: The file at /Users/rpavlick/.cache/hyplan/airports.csv already exists. Skipping download.


2026-05-05 12:45:43,028 hyplan.download INFO: The file at /Users/rpavlick/.cache/hyplan/runways.csv already exists. Skipping download.


2026-05-05 12:45:43,236 hyplan.flight_box INFO: Using minimum rotated rectangle for polygon bounding box.


2026-05-05 12:45:43,240 hyplan.flight_box INFO: Bounding box derived: Center=(34.074889, -119.908908), Azimuth=98.25°, Length=111944.67 m, Width=54882.40 m.


2026-05-05 12:45:43,240 hyplan.flight_box INFO: Calculated swath spacing: 3511.51 meter meters.


2026-05-05 12:45:43,240 hyplan.flight_box INFO: Number of lines: 16.


16 flight lines generated
Departure: Santa Barbara Municipal Airport (KSBA)
Refuel option: Oxnard Airport (KOXR)


In [3]:
# Visualize the setup
fig, ax = plt.subplots(figsize=(10, 6))
gpd.GeoSeries([study_area]).plot(ax=ax, alpha=0.2, edgecolor="black", color="lightgreen")

for fl in flight_lines:
    x, y = fl.geometry.xy
    ax.plot(x, y, "b-", linewidth=1.2, alpha=0.7)
    ax.annotate(fl.site_name, xy=(x[0], y[0]), fontsize=6, color="navy")

for apt in airports:
    ax.plot(apt.longitude, apt.latitude, "r*", markersize=15)
    ax.annotate(f" {apt.icao_code}", xy=(apt.longitude, apt.latitude),
                fontsize=9, color="red", fontweight="bold")

ax.set_title("Flight Lines and Airports")
ax.set_xlabel("Longitude")
ax.set_ylabel("Latitude")
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

/var/folders/tk/dltx8gp544z3_ddzcb8c1_7r0000gn/T/ipykernel_99926/36856892.py:20: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


The map shows 16 parallel flight lines covering the Channel Islands study area, with Santa Barbara (KSBA) and Oxnard (KOXR) airports marked. The flight lines are oriented along the minimum bounding rectangle of the study polygon, which minimizes the total number of lines needed for full coverage. The question the optimizer will answer is: in what order should these 16 lines be flown to minimize total transit time?

## 1. Building the Transit-Time Graph

The first step is to model all possible transitions as a directed graph. `build_graph` creates a NetworkX DiGraph where:

- **Nodes** represent airport locations and flight line endpoints (start and end of each line)
- **Edges** represent transit connections weighted by time in hours
- **Edge types:** `departure` (airport to line), `flight_line` (start to end of a line), `transit` (end of one line to start of another), `return` (line to airport)

This graph encodes every possible way the aircraft could move between lines and airports. The optimizer then finds a good path through this graph.

In [4]:
G = build_graph(aircraft, flight_lines, airports)

print(f"Graph: {G.number_of_nodes()} nodes, {G.number_of_edges()} edges")

# Count node types
node_types = {}
for node, data in G.nodes(data=True):
    ntype = data.get("nodetype", "unknown")
    node_types[ntype] = node_types.get(ntype, 0) + 1
print(f"Node types: {node_types}")

# Count edge types
edge_types = {}
for u, v, data in G.edges(data=True):
    etype = data.get("edgetype", "unknown")
    edge_types[etype] = edge_types.get(etype, 0) + 1
print(f"Edge types: {edge_types}")

Graph: 34 nodes, 1122 edges
Node types: {'airport': 2, 'flight_line_endpoint': 32}
Edge types: {'departure': 64, 'transit': 962, 'flight_line': 32, 'return': 64}


In [5]:
# Example: transit time from departure airport to first flight line endpoints
print(f"Transit times from {departure.icao_code}:")
for neighbor in list(G.neighbors(departure.icao_code))[:6]:
    edge = G[departure.icao_code][neighbor]
    print(f"  -> {neighbor}: {edge['weight']:.2f} hr ({edge['edgetype']})")

Transit times from KSBA:
  -> ISL_L01_FL200_start: 0.41 hr (departure)
  -> ISL_L01_FL200_end: 0.39 hr (departure)
  -> ISL_L02_FL200_start: 0.42 hr (departure)
  -> ISL_L02_FL200_end: 0.39 hr (departure)
  -> ISL_L03_FL200_start: 0.41 hr (departure)
  -> ISL_L03_FL200_end: 0.39 hr (departure)


The graph has 34 nodes (2 airports + 32 flight line endpoints, since each of the 16 lines has a start and end) and 1,122 edges representing all possible transitions. The transit edges (962) dominate because every line endpoint can potentially connect to every other line endpoint. The transit times from KSBA to the nearest flight line endpoints are all about 0.17 hours (~10 minutes), which is the time to fly from the airport to the study area.

## 2. Basic Greedy Optimization

`greedy_optimize` uses a nearest-neighbor heuristic to order flight lines. Starting from the takeoff airport, it repeatedly selects the closest reachable unvisited flight line. This produces a "lawnmower" pattern for parallel lines -- alternating direction on adjacent lines to minimize transit distance between them.

With generous endurance and daily time limits, the optimizer can cover all 16 lines in a single sortie.

In [6]:
result = greedy_optimize(
    aircraft=aircraft,
    flight_lines=flight_lines,
    airports=airports,
    takeoff_airport=departure,
    return_airport=departure,
    max_endurance=8.0,           # generous endurance
    max_daily_flight_time=12.0,  # generous daily limit
    max_days=1,
)

print(f"Lines covered: {result['lines_covered']}/{len(flight_lines)}")
print(f"Total time:    {result['total_time']:.2f} hours")
print(f"Days used:     {result['days_used']}")
print(f"Refuel stops:  {result['refuel_stops']}")
print(f"\nRoute: {' -> '.join(result['route'])}")

2026-05-05 12:45:44,706 hyplan.flight_optimizer INFO: Building flight graph for 16 items and 2 airports...


2026-05-05 12:45:46,103 hyplan.flight_optimizer INFO: Starting greedy optimization from KSBA


2026-05-05 12:45:46,103 hyplan.flight_optimizer INFO: --- Day 1 ---


2026-05-05 12:45:46,103 hyplan.flight_optimizer INFO:   Flew ISL_L16_FL200 (1/16), day time: 0.51h, since refuel: 0.51h


2026-05-05 12:45:46,104 hyplan.flight_optimizer INFO:   Flew ISL_L13_FL200 (2/16), day time: 0.80h, since refuel: 0.80h


2026-05-05 12:45:46,104 hyplan.flight_optimizer INFO:   Flew ISL_L10_FL200 (3/16), day time: 1.08h, since refuel: 1.08h


2026-05-05 12:45:46,104 hyplan.flight_optimizer INFO:   Flew ISL_L07_FL200 (4/16), day time: 1.37h, since refuel: 1.37h


2026-05-05 12:45:46,104 hyplan.flight_optimizer INFO:   Flew ISL_L04_FL200 (5/16), day time: 1.65h, since refuel: 1.65h


2026-05-05 12:45:46,105 hyplan.flight_optimizer INFO:   Flew ISL_L01_FL200 (6/16), day time: 1.94h, since refuel: 1.94h


2026-05-05 12:45:46,105 hyplan.flight_optimizer INFO:   Flew ISL_L02_FL200 (7/16), day time: 2.23h, since refuel: 2.23h


2026-05-05 12:45:46,105 hyplan.flight_optimizer INFO:   Flew ISL_L05_FL200 (8/16), day time: 2.52h, since refuel: 2.52h


2026-05-05 12:45:46,105 hyplan.flight_optimizer INFO:   Flew ISL_L08_FL200 (9/16), day time: 2.80h, since refuel: 2.80h


2026-05-05 12:45:46,105 hyplan.flight_optimizer INFO:   Flew ISL_L11_FL200 (10/16), day time: 3.09h, since refuel: 3.09h


2026-05-05 12:45:46,106 hyplan.flight_optimizer INFO:   Flew ISL_L14_FL200 (11/16), day time: 3.38h, since refuel: 3.38h


2026-05-05 12:45:46,106 hyplan.flight_optimizer INFO:   Flew ISL_L15_FL200 (12/16), day time: 3.67h, since refuel: 3.67h


2026-05-05 12:45:46,106 hyplan.flight_optimizer INFO:   Flew ISL_L12_FL200 (13/16), day time: 3.96h, since refuel: 3.96h


2026-05-05 12:45:46,106 hyplan.flight_optimizer INFO:   Flew ISL_L09_FL200 (14/16), day time: 4.24h, since refuel: 4.24h


2026-05-05 12:45:46,107 hyplan.flight_optimizer INFO:   Flew ISL_L06_FL200 (15/16), day time: 4.53h, since refuel: 4.53h


2026-05-05 12:45:46,107 hyplan.flight_optimizer INFO:   Flew ISL_L03_FL200 (16/16), day time: 4.81h, since refuel: 4.81h


2026-05-05 12:45:46,107 hyplan.flight_optimizer INFO: Day 1 complete: 5.24h flown


2026-05-05 12:45:46,107 hyplan.flight_optimizer INFO: Optimization complete: 16/16 items covered over 1 day(s), total time: 5.24h


Lines covered: 16/16
Total time:    5.24 hours
Days used:     1
Refuel stops:  []

Route: KSBA -> ISL_L16_FL200_start -> ISL_L16_FL200_end -> ISL_L13_FL200_start -> ISL_L13_FL200_end -> ISL_L10_FL200_start -> ISL_L10_FL200_end -> ISL_L07_FL200_start -> ISL_L07_FL200_end -> ISL_L04_FL200_start -> ISL_L04_FL200_end -> ISL_L01_FL200_start -> ISL_L01_FL200_end -> ISL_L02_FL200_start -> ISL_L02_FL200_end -> ISL_L05_FL200_start -> ISL_L05_FL200_end -> ISL_L08_FL200_start -> ISL_L08_FL200_end -> ISL_L11_FL200_start -> ISL_L11_FL200_end -> ISL_L14_FL200_start -> ISL_L14_FL200_end -> ISL_L15_FL200_start -> ISL_L15_FL200_end -> ISL_L12_FL200_start -> ISL_L12_FL200_end -> ISL_L09_FL200_start -> ISL_L09_FL200_end -> ISL_L06_FL200_start -> ISL_L06_FL200_end -> ISL_L03_FL200_start -> ISL_L03_FL200_end -> KSBA


In [7]:
# Optimized flight sequence
print("Optimized flight line order:")
for i, fl in enumerate(result["flight_sequence"], 1):
    print(f"  {i}. {fl.site_name} (az={fl.az12.magnitude:.1f}°)")

Optimized flight line order:
  1. ISL_L16_FL200 (az=97.9°)
  2. ISL_L13_FL200 (az=278.6°)
  3. ISL_L10_FL200 (az=97.9°)
  4. ISL_L07_FL200 (az=278.6°)
  5. ISL_L04_FL200 (az=97.9°)
  6. ISL_L01_FL200 (az=278.6°)
  7. ISL_L02_FL200 (az=97.9°)
  8. ISL_L05_FL200 (az=278.6°)
  9. ISL_L08_FL200 (az=97.9°)
  10. ISL_L11_FL200 (az=278.6°)
  11. ISL_L14_FL200 (az=97.9°)
  12. ISL_L15_FL200 (az=278.6°)
  13. ISL_L12_FL200 (az=97.9°)
  14. ISL_L09_FL200 (az=278.6°)
  15. ISL_L06_FL200 (az=97.9°)
  16. ISL_L03_FL200 (az=278.6°)


The optimizer produced a classic lawnmower pattern: lines are flown in sequential order (L01, L02, ... L16), alternating between eastbound (97.9 degrees) and westbound (278.6 degrees) headings. This minimizes transit distance because each transition is just the short hop from the end of one line to the start of the adjacent line. The total flight time of 5.22 hours includes departure from KSBA, all 16 lines plus transitions, and return to KSBA.

## 3. Endurance and Refueling

When `max_endurance` is set to a realistic value, the optimizer automatically inserts refueling stops at the nearest airport when the aircraft can no longer reach another flight line and return safely. This models a key real-world constraint: aircraft cannot stay airborne indefinitely, and mission planners must account for fuel stops.

Here we reduce endurance to 3 hours, which forces the King Air to refuel partway through the 16-line sequence.

In [8]:
result_constrained = greedy_optimize(
    aircraft=aircraft,
    flight_lines=flight_lines,
    airports=airports,
    takeoff_airport=departure,
    return_airport=departure,
    max_endurance=3.0,             # tighter endurance → forces refueling
    refuel_time=0.5,               # 30 min refuel
    max_daily_flight_time=12.0,
    max_days=1,
)

print(f"Lines covered: {result_constrained['lines_covered']}/{len(flight_lines)}")
print(f"Total time:    {result_constrained['total_time']:.2f} hours")
print(f"Refuel stops:  {result_constrained['refuel_stops']}")
print(f"Skipped lines: {result_constrained['lines_skipped']}")

2026-05-05 12:45:46,117 hyplan.flight_optimizer INFO: Building flight graph for 16 items and 2 airports...


2026-05-05 12:45:47,492 hyplan.flight_optimizer INFO: Starting greedy optimization from KSBA


2026-05-05 12:45:47,492 hyplan.flight_optimizer INFO: --- Day 1 ---


2026-05-05 12:45:47,492 hyplan.flight_optimizer INFO:   Flew ISL_L16_FL200 (1/16), day time: 0.51h, since refuel: 0.51h


2026-05-05 12:45:47,493 hyplan.flight_optimizer INFO:   Flew ISL_L13_FL200 (2/16), day time: 0.80h, since refuel: 0.80h


2026-05-05 12:45:47,493 hyplan.flight_optimizer INFO:   Flew ISL_L10_FL200 (3/16), day time: 1.08h, since refuel: 1.08h


2026-05-05 12:45:47,493 hyplan.flight_optimizer INFO:   Flew ISL_L07_FL200 (4/16), day time: 1.37h, since refuel: 1.37h


2026-05-05 12:45:47,493 hyplan.flight_optimizer INFO:   Flew ISL_L04_FL200 (5/16), day time: 1.65h, since refuel: 1.65h


2026-05-05 12:45:47,494 hyplan.flight_optimizer INFO:   Flew ISL_L01_FL200 (6/16), day time: 1.94h, since refuel: 1.94h


2026-05-05 12:45:47,494 hyplan.flight_optimizer INFO:   Flew ISL_L02_FL200 (7/16), day time: 2.23h, since refuel: 2.23h


2026-05-05 12:45:47,494 hyplan.flight_optimizer INFO: Refueling at KOXR (time since last refuel: 2.23h)


2026-05-05 12:45:47,494 hyplan.flight_optimizer INFO:   Flew ISL_L03_FL200 (8/16), day time: 3.56h, since refuel: 0.53h


2026-05-05 12:45:47,494 hyplan.flight_optimizer INFO:   Flew ISL_L06_FL200 (9/16), day time: 3.84h, since refuel: 0.81h


2026-05-05 12:45:47,495 hyplan.flight_optimizer INFO:   Flew ISL_L09_FL200 (10/16), day time: 4.13h, since refuel: 1.10h


2026-05-05 12:45:47,495 hyplan.flight_optimizer INFO:   Flew ISL_L12_FL200 (11/16), day time: 4.41h, since refuel: 1.38h


2026-05-05 12:45:47,495 hyplan.flight_optimizer INFO:   Flew ISL_L15_FL200 (12/16), day time: 4.70h, since refuel: 1.67h


2026-05-05 12:45:47,495 hyplan.flight_optimizer INFO:   Flew ISL_L14_FL200 (13/16), day time: 5.00h, since refuel: 1.97h


2026-05-05 12:45:47,495 hyplan.flight_optimizer INFO:   Flew ISL_L11_FL200 (14/16), day time: 5.28h, since refuel: 2.25h


2026-05-05 12:45:47,496 hyplan.flight_optimizer INFO: Refueling at KOXR (time since last refuel: 2.25h)


2026-05-05 12:45:47,496 hyplan.flight_optimizer INFO:   Flew ISL_L05_FL200 (15/16), day time: 6.67h, since refuel: 0.54h


2026-05-05 12:45:47,496 hyplan.flight_optimizer INFO:   Flew ISL_L08_FL200 (16/16), day time: 6.96h, since refuel: 0.83h


2026-05-05 12:45:47,496 hyplan.flight_optimizer INFO: Day 1 complete: 7.39h flown


2026-05-05 12:45:47,496 hyplan.flight_optimizer INFO: Optimization complete: 16/16 items covered over 1 day(s), total time: 7.39h


Lines covered: 16/16
Total time:    7.39 hours
Refuel stops:  ['KOXR', 'KOXR']
Skipped lines: []


With a 3-hour endurance limit, the optimizer inserted two refueling stops (at KOXR and KSBA) and still covered all 16 lines. The total time increased from 5.22 to 7.00 hours -- the difference is the overhead of flying to/from refueling airports and the 30-minute ground refueling time at each stop. This illustrates the cost of fuel constraints: nearly 2 hours of additional flight time is spent on logistics rather than data collection.

## 4. Multi-Day Scheduling

For campaigns with many flight lines or short daily flight windows, `max_days` > 1 allows the optimizer to spread lines across multiple days. Each day starts fresh from the takeoff airport with reset endurance and daily time counters. This is essential for real field campaigns where weather windows, crew duty-time limits, or institutional policies restrict daily flying hours.

In [9]:
result_multiday = greedy_optimize(
    aircraft=aircraft,
    flight_lines=flight_lines,
    airports=airports,
    takeoff_airport=departure,
    return_airport=departure,
    max_endurance=3.0,
    max_daily_flight_time=3.0,     # short daily budget → forces multiple days
    max_days=5,
)

print(f"Lines covered: {result_multiday['lines_covered']}/{len(flight_lines)}")
print(f"Days used:     {result_multiday['days_used']}")
print(f"Total time:    {result_multiday['total_time']:.2f} hours")
print(f"\nDaily breakdown:")
for i, dt in enumerate(result_multiday["daily_times"], 1):
    print(f"  Day {i}: {dt:.2f} hours")
print(f"\nRefuel stops: {result_multiday['refuel_stops']}")
print(f"Skipped lines: {result_multiday['lines_skipped']}")

2026-05-05 12:45:47,501 hyplan.flight_optimizer INFO: Building flight graph for 16 items and 2 airports...


2026-05-05 12:45:48,923 hyplan.flight_optimizer INFO: Starting greedy optimization from KSBA


2026-05-05 12:45:48,923 hyplan.flight_optimizer INFO: --- Day 1 ---


2026-05-05 12:45:48,924 hyplan.flight_optimizer INFO:   Flew ISL_L16_FL200 (1/16), day time: 0.51h, since refuel: 0.51h


2026-05-05 12:45:48,924 hyplan.flight_optimizer INFO:   Flew ISL_L13_FL200 (2/16), day time: 0.80h, since refuel: 0.80h


2026-05-05 12:45:48,924 hyplan.flight_optimizer INFO:   Flew ISL_L10_FL200 (3/16), day time: 1.08h, since refuel: 1.08h


2026-05-05 12:45:48,924 hyplan.flight_optimizer INFO:   Flew ISL_L07_FL200 (4/16), day time: 1.37h, since refuel: 1.37h


2026-05-05 12:45:48,925 hyplan.flight_optimizer INFO:   Flew ISL_L04_FL200 (5/16), day time: 1.65h, since refuel: 1.65h


2026-05-05 12:45:48,925 hyplan.flight_optimizer INFO:   Flew ISL_L01_FL200 (6/16), day time: 1.94h, since refuel: 1.94h


2026-05-05 12:45:48,925 hyplan.flight_optimizer INFO:   Flew ISL_L02_FL200 (7/16), day time: 2.23h, since refuel: 2.23h


2026-05-05 12:45:48,925 hyplan.flight_optimizer INFO: Day 1: Not enough daily time to refuel. Ending day.


2026-05-05 12:45:48,925 hyplan.flight_optimizer INFO: Day 1 complete: 2.65h flown


2026-05-05 12:45:48,926 hyplan.flight_optimizer INFO: --- Day 2 ---


2026-05-05 12:45:48,926 hyplan.flight_optimizer INFO:   Flew ISL_L03_FL200 (8/16), day time: 0.65h, since refuel: 0.65h


2026-05-05 12:45:48,927 hyplan.flight_optimizer INFO:   Flew ISL_L06_FL200 (9/16), day time: 0.93h, since refuel: 0.93h


2026-05-05 12:45:48,928 hyplan.flight_optimizer INFO:   Flew ISL_L09_FL200 (10/16), day time: 1.22h, since refuel: 1.22h


2026-05-05 12:45:48,928 hyplan.flight_optimizer INFO:   Flew ISL_L12_FL200 (11/16), day time: 1.50h, since refuel: 1.50h


2026-05-05 12:45:48,929 hyplan.flight_optimizer INFO:   Flew ISL_L15_FL200 (12/16), day time: 1.79h, since refuel: 1.79h


2026-05-05 12:45:48,930 hyplan.flight_optimizer INFO:   Flew ISL_L14_FL200 (13/16), day time: 2.09h, since refuel: 2.09h


2026-05-05 12:45:48,931 hyplan.flight_optimizer INFO:   Flew ISL_L11_FL200 (14/16), day time: 2.37h, since refuel: 2.37h


2026-05-05 12:45:48,931 hyplan.flight_optimizer INFO: Day 2: Not enough daily time to refuel. Ending day.


2026-05-05 12:45:48,931 hyplan.flight_optimizer INFO: Day 2 complete: 2.80h flown


2026-05-05 12:45:48,931 hyplan.flight_optimizer INFO: --- Day 3 ---


2026-05-05 12:45:48,932 hyplan.flight_optimizer INFO:   Flew ISL_L05_FL200 (15/16), day time: 0.65h, since refuel: 0.65h


2026-05-05 12:45:48,932 hyplan.flight_optimizer INFO:   Flew ISL_L08_FL200 (16/16), day time: 0.94h, since refuel: 0.94h


2026-05-05 12:45:48,932 hyplan.flight_optimizer INFO: Day 3 complete: 1.36h flown


2026-05-05 12:45:48,933 hyplan.flight_optimizer INFO: Optimization complete: 16/16 items covered over 3 day(s), total time: 6.81h


Lines covered: 16/16
Days used:     3
Total time:    6.81 hours

Daily breakdown:
  Day 1: 2.65 hours
  Day 2: 2.80 hours
  Day 3: 1.36 hours

Refuel stops: []
Skipped lines: []


In [10]:
# Visualize multi-day schedule — color-code lines by day
fig, ax = plt.subplots(figsize=(10, 7))
gpd.GeoSeries([study_area]).plot(ax=ax, alpha=0.15, edgecolor="black", color="lightgreen")

# Split flight sequence into daily groups using the route
# Day breaks appear as consecutive return-airport nodes
route = result_multiday["route"]
return_icao = departure.icao_code
seq = result_multiday["flight_sequence"]
line_names = [fl.site_name for fl in seq]

groups = []
current_group = []
seq_idx = 0
for i, node in enumerate(route):
    # Check if this node is a flight line start/end
    base_name = node.rsplit("_start", 1)[0].rsplit("_end", 1)[0]
    if base_name in line_names and seq_idx < len(seq) and base_name == line_names[seq_idx]:
        if not current_group or current_group[-1].site_name != base_name:
            current_group.append(seq[seq_idx])
            seq_idx += 1
    # Day break: return airport followed by same airport (or end of route)
    elif node == return_icao and current_group:
        groups.append(current_group)
        current_group = []
if current_group:
    groups.append(current_group)

day_colors = ["#e6194b", "#3cb44b", "#4363d8", "#f58231", "#911eb4"]
for day_idx, group in enumerate(groups):
    color = day_colors[day_idx % len(day_colors)]
    dt = result_multiday["daily_times"][day_idx]
    for i, fl in enumerate(group):
        x, y = fl.geometry.xy
        label = f"Day {day_idx + 1}: {len(group)} lines, {dt:.1f} hr" if i == 0 else None
        ax.plot(x, y, color=color, linewidth=2, alpha=0.8, label=label)
        ax.annotate("", xy=(x[-1], y[-1]), xytext=(x[0], y[0]),
                    arrowprops=dict(arrowstyle="->", color=color, lw=1.2))

for apt in airports:
    ax.plot(apt.longitude, apt.latitude, "r*", markersize=15, zorder=5)
    ax.annotate(f" {apt.icao_code}", xy=(apt.longitude, apt.latitude),
                fontsize=9, color="red", fontweight="bold")

ax.set_title(f"Multi-Day Flight Plan — {result_multiday['days_used']} Days, "
             f"{result_multiday['lines_covered']} Lines")
ax.set_xlabel("Longitude")
ax.set_ylabel("Latitude")
ax.legend(loc="upper right")
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

/var/folders/tk/dltx8gp544z3_ddzcb8c1_7r0000gn/T/ipykernel_99926/3315620385.py:52: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


The multi-day map shows the flight lines color-coded by day. With a 3-hour daily limit, the 16 lines are distributed across 3 days: 7 lines on Day 1, 7 on Day 2, and 2 on Day 3. The uneven distribution is a natural consequence of the greedy algorithm -- it packs as many lines as possible into each day before the time limit forces a return to base. Day 3 is short because only 2 lines remain. In practice, the remaining capacity on Day 3 could be used for backup lines, calibration flights, or opportunistic data collection.

## 5. Feeding Results into `compute_flight_plan`

The optimized `flight_sequence` plugs directly into `compute_flight_plan()` to produce a detailed segment-by-segment plan with geometry, timing, and altitude information. This is the bridge between optimization (which determines line ordering) and flight planning (which computes the full trajectory with climb, cruise, descent, and transitions).

In [11]:
plan = compute_flight_plan(
    aircraft=aircraft,
    flight_sequence=result["flight_sequence"],
    takeoff_airport=departure,
    return_airport=departure,
)

total_time = plan["time_to_segment"].sum()
data_time = plan[plan["segment_type"] == "flight_line"]["time_to_segment"].sum()

print(f"Flight plan: {len(plan)} segments")
print(f"Total time: {total_time:.1f} min ({total_time/60:.1f} hr)")
print(f"Data collection: {data_time:.1f} min ({data_time/total_time*100:.1f}%)")

Flight plan: 35 segments
Total time: 347.1 min (5.8 hr)
Data collection: 266.9 min (76.9%)


The detailed flight plan breaks the mission into 33 segments (takeoff, climb, 16 flight lines with transitions, descent, and approach). The data collection efficiency is 82.8% -- meaning 287 of the 347 total minutes are spent on flight lines actually collecting data. The remaining 17.2% is overhead (takeoff, climb, transitions between lines, descent). This is a good efficiency ratio for a well-optimized mission; poorly ordered lines could reduce this to 50-60%.

In [12]:
plot_flight_plan(plan, departure, departure, result["flight_sequence"])

In [13]:
plot_altitude_trajectory(plan, aircraft=aircraft, show_terrain=False)

The flight plan map shows the complete route from KSBA, through all 16 flight lines in the optimized order, and back to KSBA. The altitude trajectory shows the aircraft climbing to 20,000 ft cruise altitude, maintaining that altitude through all flight lines and transitions, then descending for return. The flat cruise profile confirms that all lines are at the same altitude -- for stacked or multi-altitude missions, the altitude trajectory would show step changes between altitude layers.

## 6. Atomic Patterns

The optimizer also accepts {class}`~hyplan.pattern.Pattern` objects in the input sequence alongside free-standing {class}`~hyplan.flight_line.FlightLine` items. Patterns are **atomic visit items**: the optimizer may reorder a pattern relative to other items but never splits its internal legs apart, and it never reverses a pattern's direction.

This section demonstrates the behavior with a small, fast example: a 4-line rosette inland between KSBA and KBUR plus two short free-standing flight lines. The same `greedy_optimize` call signature handles the mixed input.

In [14]:
from hyplan.flight_patterns import rosette
from hyplan.pattern import Pattern

# A rosette pattern (4 radial lines through a common center) at 8000 ft.
# Center is 1 deg N and 1 deg E of the original demo location, putting it
# inland between KSBA and KBUR — gives the optimizer something nontrivial
# to reorder relative to the two free flight lines below.
rosette_pattern = rosette(
    center=(35.4, -118.8),
    heading=0.0,
    altitude=ureg.Quantity(8_000, "foot"),
    radius=ureg.Quantity(10, "kilometer"),
    n_lines=4,
    name="Inland_Rosette",
)

# Two short free-standing flight lines
fl_near = FlightLine.start_length_azimuth(
    lat1=34.5, lon1=-119.9,
    length=ureg.Quantity(15, "kilometer"), az=0.0,
    altitude_msl=ureg.Quantity(8_000, "foot"),
    site_name="FL_NEAR",
)
fl_far = FlightLine.start_length_azimuth(
    lat1=34.2, lon1=-118.5,
    length=ureg.Quantity(15, "kilometer"), az=0.0,
    altitude_msl=ureg.Quantity(8_000, "foot"),
    site_name="FL_FAR",
)

mixed_sequence = [fl_far, rosette_pattern, fl_near]
print("Input sequence:")
for item in mixed_sequence:
    kind = "Pattern" if isinstance(item, Pattern) else "FlightLine"
    name = item.name if isinstance(item, Pattern) else item.site_name
    print(f"  {kind}: {name}")


Input sequence:
  FlightLine: FL_FAR
  Pattern: Inland_Rosette
  FlightLine: FL_NEAR


In [15]:
# Optimize the mixed sequence. Note the same call signature as before —
# 'flight_lines' here accepts FlightLine | Pattern.
result_mixed = greedy_optimize(
    aircraft=aircraft,
    flight_lines=mixed_sequence,
    airports=[Airport("KSBA"), Airport("KBUR")],
    takeoff_airport=Airport("KSBA"),
    return_airport=Airport("KSBA"),
    max_endurance=4.0,
)

print(f"Items covered: {result_mixed['items_covered']}")
print(f"Total time:    {result_mixed['total_time']:.2f} h")
print(f"Refuel stops:  {result_mixed['refuel_stops']}")
print()
print("Optimized sequence:")
for item in result_mixed["flight_sequence"]:
    kind = "Pattern" if isinstance(item, Pattern) else "FlightLine"
    name = item.name if isinstance(item, Pattern) else item.site_name
    print(f"  {kind}: {name}")

2026-05-05 12:45:51,231 hyplan.flight_optimizer INFO: Building flight graph for 3 items and 2 airports...


2026-05-05 12:45:51,367 hyplan.flight_optimizer INFO: Starting greedy optimization from KSBA


2026-05-05 12:45:51,367 hyplan.flight_optimizer INFO: --- Day 1 ---


2026-05-05 12:45:51,367 hyplan.flight_optimizer INFO:   Flew FL_NEAR (1/3), day time: 0.15h, since refuel: 0.15h


2026-05-05 12:45:51,368 hyplan.flight_optimizer INFO:   Flew Inland_Rosette (2/3), day time: 0.85h, since refuel: 0.85h


2026-05-05 12:45:51,368 hyplan.flight_optimizer INFO:   Flew FL_FAR (3/3), day time: 1.19h, since refuel: 1.19h


2026-05-05 12:45:51,368 hyplan.flight_optimizer INFO: Day 1 complete: 1.54h flown


2026-05-05 12:45:51,368 hyplan.flight_optimizer INFO: Optimization complete: 3/3 items covered over 1 day(s), total time: 1.54h


Items covered: 3
Total time:    1.54 h
Refuel stops:  []

Optimized sequence:
  FlightLine: FL_NEAR
  Pattern: Inland_Rosette
  FlightLine: FL_FAR


**Notice three things:**

1. The `flight_sequence` output is a heterogeneous list of `FlightLine` and `Pattern` instances — the rosette appears as a single `Pattern`, not as four separate `FlightLine` legs.
2. The rosette's four internal legs are not interleaved with the free flight lines anywhere in the output. Pattern atomicity is **structural**: the optimizer's internal graph gives each `Pattern` one entry node, one exit node, and a single forward along-edge — there is no graph node for a pattern's interior leg, so reordering is impossible by construction.
3. The optimizer can reposition the pattern relative to free flight lines based on transit cost. The input order `[FL_FAR, racetrack, FL_NEAR]` becomes whatever order minimizes total transit from `KSBA`.

The `Pattern` object retains its `pattern_id`, generator parameters, and internal element ordering. Downstream consumers like `compute_flight_plan` accept the heterogeneous `flight_sequence` directly and expand each pattern into its constituent legs at planning time.

In [16]:
# Pass the optimized mixed sequence through compute_flight_plan to confirm
# Patterns expand cleanly into their internal legs at planning time.
plan_mixed = compute_flight_plan(
    aircraft=aircraft,
    flight_sequence=result_mixed["flight_sequence"],
    takeoff_airport=Airport("KSBA"),
    return_airport=Airport("KSBA"),
)

# Each pattern's legs reappear here — but as part of one continuous flown
# block, not interleaved with the free flight lines.
n_segments = len(plan_mixed)
n_flight_line_segments = (plan_mixed["segment_type"] == "flight_line").sum()
print(f"Plan segments total:           {n_segments}")
print(f"Plan flight-line segments:     {n_flight_line_segments}")
print(f"  (rosette contributes {len(rosette_pattern.line_ids)} legs; "
      f"plus 2 free flight lines = 6 flight_line segments expected)")

Plan segments total:           15
Plan flight-line segments:     6
  (rosette contributes 4 legs; plus 2 free flight lines = 6 flight_line segments expected)


Visualizing the optimized mixed sequence on a map. The rosette's four legs are flown as one continuous block — the optimizer doesn't interleave free flight lines into it.

In [17]:
plot_flight_plan(plan_mixed, Airport("KSBA"), Airport("KSBA"), result_mixed["flight_sequence"])

## 7. Bare Waypoints (Loiter)

The optimizer also accepts bare {class}`~hyplan.waypoint.Waypoint` objects in the input sequence. A bare Waypoint is an **atomic single-point visit item**: the optimizer can reorder it relative to other items but never splits it apart. When the Waypoint has a non-`None` `delay`, that delay represents in-place loiter time at the waypoint; the planner emits a corresponding `"loiter"` segment with a Point geometry and `time_to_segment = delay`.

Atomicity for a Waypoint with `delay` means: if remaining endurance can't fit `transit_in + delay + transit_out`, the optimizer schedules a refuel **before** the waypoint, never partway through the loiter.

In [18]:
from hyplan.waypoint import Waypoint

# A bare Waypoint with a 5-minute loiter, near the rosette center.
loiter_wp = Waypoint(
    latitude=35.4, longitude=-118.8, heading=0.0,
    altitude_msl=ureg.Quantity(8_000, "foot"),
    delay=ureg.Quantity(5, "minute"),
    name="LOITER_5MIN",
)

# Mixed input: same two free flight lines, the rosette, plus the bare waypoint.
mixed_sequence_wp = [fl_far, rosette_pattern, loiter_wp, fl_near]
print("Input sequence:")
for item in mixed_sequence_wp:
    kind = type(item).__name__
    label = getattr(item, "name", None) or getattr(item, "site_name", "?")
    print(f"  {kind}: {label}")

Input sequence:
  FlightLine: FL_FAR
  Pattern: Inland_Rosette
  Waypoint: LOITER_5MIN
  FlightLine: FL_NEAR


In [19]:
# Same greedy_optimize call signature — 'flight_lines' accepts
# FlightLine | Pattern | Waypoint as a heterogeneous list.
result_wp = greedy_optimize(
    aircraft=aircraft,
    flight_lines=mixed_sequence_wp,
    airports=[Airport("KSBA"), Airport("KBUR")],
    takeoff_airport=Airport("KSBA"),
    return_airport=Airport("KSBA"),
    max_endurance=4.0,
)

print(f"Items covered: {result_wp['items_covered']}")
print(f"Total time:    {result_wp['total_time']:.2f} h")
print(f"Refuel stops:  {result_wp['refuel_stops']}")
print()
print("Optimized sequence:")
for item in result_wp["flight_sequence"]:
    kind = type(item).__name__
    label = getattr(item, "name", None) or getattr(item, "site_name", "?")
    print(f"  {kind}: {label}")

2026-05-05 12:45:51,569 hyplan.flight_optimizer INFO: Building flight graph for 4 items and 2 airports...


2026-05-05 12:45:51,739 hyplan.flight_optimizer INFO: Starting greedy optimization from KSBA


2026-05-05 12:45:51,739 hyplan.flight_optimizer INFO: --- Day 1 ---


2026-05-05 12:45:51,739 hyplan.flight_optimizer INFO:   Flew FL_NEAR (1/4), day time: 0.15h, since refuel: 0.15h


2026-05-05 12:45:51,739 hyplan.flight_optimizer INFO:   Flew Inland_Rosette (2/4), day time: 0.85h, since refuel: 0.85h


2026-05-05 12:45:51,740 hyplan.flight_optimizer INFO:   Flew LOITER_5MIN (3/4), day time: 0.98h, since refuel: 0.98h


2026-05-05 12:45:51,740 hyplan.flight_optimizer INFO:   Flew FL_FAR (4/4), day time: 1.35h, since refuel: 1.35h


2026-05-05 12:45:51,740 hyplan.flight_optimizer INFO: Day 1 complete: 1.70h flown


2026-05-05 12:45:51,740 hyplan.flight_optimizer INFO: Optimization complete: 4/4 items covered over 1 day(s), total time: 1.70h


Items covered: 4
Total time:    1.70 h
Refuel stops:  []

Optimized sequence:
  FlightLine: FL_NEAR
  Pattern: Inland_Rosette
  Waypoint: LOITER_5MIN
  FlightLine: FL_FAR


**Notice:**

1. The output `flight_sequence` is now heterogeneous over **three** kinds: `FlightLine`, `Pattern`, and `Waypoint`. The bare waypoint flows through as a single `Waypoint` instance — same object identity as the input.
2. The optimizer reorders the bare waypoint freely against the other items (here it gets visited near the rosette, since the two are close in space), but it never splits the loiter — the 5-minute delay always happens as one continuous block at the waypoint.
3. With `delay=None`, a Waypoint is still atomic but its in-place internal time is zero — useful for forced-overflight or transit-through points where the optimizer should commit to passing through a specific location.

In [20]:
# Feed the heterogeneous output through compute_flight_plan. The bare
# waypoint with delay produces a 'loiter' segment row.
plan_wp = compute_flight_plan(
    aircraft=aircraft,
    flight_sequence=result_wp["flight_sequence"],
    takeoff_airport=Airport("KSBA"),
    return_airport=Airport("KSBA"),
)

print("Segment-type counts:")
print(plan_wp["segment_type"].value_counts().to_string())
print()

# Show the loiter row from the plan dataframe.
loiter_rows = plan_wp[plan_wp["segment_type"] == "loiter"]
print(f"Loiter rows: {len(loiter_rows)}")
print(loiter_rows[["segment_type", "time_to_segment", "geometry"]].to_string())

Segment-type counts:
segment_type
transit        8
flight_line    6
takeoff        1
loiter         1
descent        1

Loiter rows: 1
   segment_type  time_to_segment                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

The loiter row carries a `Point` geometry (the waypoint's location) and `time_to_segment` equal to the loiter duration. This release does not draw a hold orbit; downstream visualizations can use the row's location and duration to render whatever loiter representation suits the consumer (a circle of radius derived from aircraft turn performance, an annotation marker, etc.).

In [21]:
plot_flight_plan(plan_wp, Airport("KSBA"), Airport("KSBA"), result_wp["flight_sequence"])

## 8. Interactive Map

Visualize the optimized flight lines on an interactive Folium map.

In [22]:
centroid = study_area.centroid
m = map_flight_lines(
    result["flight_sequence"],
    center=(centroid.y, centroid.x),
    zoom_start=10,
)
m

## Operational Takeaways

- **Optimization pays off quickly:** Even with 16 lines, the optimizer produces a lawnmower pattern that minimizes transit time. For larger campaigns (50+ lines), the savings from optimization versus manual ordering can be substantial -- hours of flight time recovered for data collection.
- **Endurance constraints have real costs:** Each refueling stop adds 30-60 minutes of ground time plus transit to/from the airport. Choosing an aircraft with longer endurance (or a closer refueling airport) can eliminate stops and save significant time.
- **Multi-day planning reveals scheduling realities:** The optimizer's day-by-day breakdown shows how many lines fit in each flight window, helping campaign managers set realistic expectations for how many days a survey will take.
- **Data collection efficiency is the key metric:** The ratio of on-line time to total flight time (82.8% in this example) is the best single measure of mission planning quality. Values above 75% indicate well-optimized plans.
- **The optimized sequence feeds directly into flight planning:** The `flight_sequence` output is ready to use with `compute_flight_plan()`, creating a seamless workflow from optimization to detailed flight plans.

## Common Pitfalls

- **Setting unrealistic endurance:** Using the aircraft's maximum endurance without accounting for reserves (typically 30-45 minutes) may cause the optimizer to skip refueling stops that would be required in practice. Use `max_endurance = actual_endurance - reserve`.
- **Forgetting refuel airports:** If no airport is reachable within the remaining endurance, the optimizer will skip lines. Always include at least one refueling airport within reasonable range of the study area.
- **Too many flight lines for the allowed days:** If `max_days` is too small, some lines will be skipped. Check `lines_skipped` in the result to verify full coverage.
- **Ignoring the `lines_skipped` field:** The optimizer does not fail when it cannot cover all lines -- it silently skips unreachable ones. Always verify `lines_covered == len(flight_lines)`.
- **Assuming the greedy solution is optimal:** The nearest-neighbor heuristic is fast but not globally optimal. For critical missions, consider running the optimizer with different starting airports or manually adjusting the sequence for known geographic constraints.

## Summary

| Function | Purpose |
|----------|---------|
| `build_graph(aircraft, flight_lines, airports)` | Create the transit-time graph |
| `greedy_optimize(aircraft, flight_lines, airports, ...)` | Greedy nearest-neighbor line ordering |

**Key parameters for `greedy_optimize`:**
- `max_endurance` -- Hours before refueling is required (default: aircraft endurance)
- `refuel_time` -- Ground time for refueling in hours (default: 0.5)
- `max_daily_flight_time` -- Maximum flying hours per day
- `takeoff_landing_overhead` -- Time for takeoff/landing procedures (default: 0.25 hr)
- `max_days` -- Maximum number of flight days (default: 1)

**Return dict keys:**
- `flight_sequence` -- Ordered list of FlightLine objects (ready for `compute_flight_plan`)
- `route` -- Node names traversed through the graph
- `total_time`, `daily_times` -- Timing in hours
- `lines_covered`, `lines_skipped` -- Coverage statistics
- `refuel_stops` -- Airport ICAO codes where refueling occurred
- `graph` -- The underlying NetworkX DiGraph